tinylfu/tinylfu.go

```diff
@@ -0,0 +1,122 @@
+// Package tinylfu is an implementation of the W-TinyLFU caching algorithm.
+// See details at http://arxiv.org/abs/1512.00727
+package tinylfu
+
+// Policy implements a windowed TinyLFU eviction policy. It is not safe for concurrent access.
+type Policy struct {
+	data     map[uint64]*element
+	admittor AdmissionPolicy
+	stats    StatsRecorder
+
+	window    *list
+	probation *list
+	protected *list
+
+	capacity     int
+	maxWindow    int
+	maxProtected int
+}
+
+// New creates a new TinyLFU cache.
+func New(capacity int, opts ...Option) *Policy {
+	// Consistent behavior relies on capacity for one element in each segment.
+	if capacity < 3 {
+		panic("tinylfu: capacity must be at least 3 (1 for each segment)")
+	}
+
+	p := &Policy{
+		data:      make(map[uint64]*element),
+		window:    newList(),
+		probation: newList(),
+		protected: newList(),
+		capacity:  capacity,
+	}
+
+	WithSegmentation(0.99, 0.8)(p)
+	for _, opt := range opts {
+		opt(p)
+	}
+	return p
+}
+
+// Len returns the number of items in the cache.
+func (p *Policy) Len() int {
+	return p.window.Len() + p.probation.Len() + p.protected.Len()
+}
+
+// Record updates the policy when an entry is accessed.
+func (p *Policy) Record(key uint64) {
+	if p.admittor != nil {
+		p.admittor.Record(key)
+	}
+
+	node, ok := p.data[key]
+	if !ok {
+		if p.stats != nil {
```

Comment:
Instead of checking this everywhere, we can set the stats to a default StatsRecorder, which could be something meaningful or just could be a noop. Same for AdmissionPolicy.